engine/consensus/matching/engine.go

```diff
@@ -158,6 +233,35 @@ func (e *Engine) loop() {
 	}
 }
 
+// processLatestFinalizedEvent performs processing of latest finalized event propagating it to core
+func (e *Engine) processLatestFinalizedEvent() error {
+	err := e.core.OnBlockFinalization()
+	if err != nil {
+		return fmt.Errorf("could not process last finalized event: %w", err)
+	}
+	return nil
+}
+
+// processBlockIncorporatedEvents performs processing of block incorporated hot stuff events
+func (e *Engine) processBlockIncorporatedEvents() error {
+	for {
+		select {
+		case <-e.unit.Quit():
+			return nil
+		default:
+		}
+
+		msg, ok := e.pendingIncorporatedBlocks.Pop()
+		if ok {
+			err := e.processIncorporatedBlock(msg.(flow.Identifier))
+			if err != nil {
+				return fmt.Errorf("could not process incorporated block: %w", err)
+			}
+			continue
+		}
+	}
```

Comment:
⚠️ I think you are missing a `return` statement here:
 ```suggestion
		return nil // no more events to process
	}
```
Otherwise, you'll never get out of this loop again.